src/Microsoft.Kusto.ServiceLayer/DataSource/KustoClient.cs

```diff
@@ -183,36 +189,40 @@ public IDataReader ExecuteQuery(string query, CancellationToken cancellationToke
 
                     origReaders[index] = origReader;
                 });
-                
+
                 return new KustoResultsReader(origReaders);
             }
-            catch (AggregateException exception) 
-                when (retryCount > 0 &&
-                      exception.InnerException is KustoRequestException innerException 
+            catch (AggregateException exception)
+                when (retryCount < RetryLimit &&
```

Comment:
I think we should use the RetryPolicy which is setup for the ReliableDataSourceConnection. Maybe pass it in here. It can handle the retries, etc. This way the user can override our policy. If the user does not pass in a retry policy we can enforce our default by creating a RetryPolicy which reflect what you are trying to do here.

Reply:
There are various types of retry policies defined in RetryPolicy.cs, including ExponentialDelayRetryPolicy.